### PR TITLE
Y.F.C Buffer Cache Revamp

### PR DESCRIPTION
**Pull request by FernandoS27**
This PR introduces a overhaul of a lot of the systems from the buffer cache. It brings a set of fixes and introduces async buffer downloads.

Fixes:
- Many games that required High to be visually accurate, now work perfectly in Normal with minimal sacrifice.
- Particles in Zelda Breath of the Wild have been fixed, resulting in higher performance with accurate rendering at Normal GPU accuracy
- Tree Flickering in Zelda Breath of the Wild has been fixed on all GPU accuracies
- Particles & Character lighting/shading in Pokemon Sword & Shield has been fixed at Normal GPU accuracy. Performance & rendering is now best best by up to 40% at Normal accuracy
- Models (BowWow for example) & particle rendering fixed at Normal GPU accuracy in Zelda Links Awakening. Performance at Normal accuracy, with correct rendering is now up to 70% higher than before (When accurately rendered)
- Lighting in Diablo 2 Resurrected has been fixed and will no longer flicker.
- Lighting and Shadows in Luigi's Mansion 3 have been fixed and will no longer randomly flicker
- Pokemon photograph detection and data of Pokemon Snap has been fixed art normal GPU accuracy. This results in up to a 50% increase to performance with working photograph detection
- Kirby and the Forgotten Land vertex explosions,  lighting & particles have been fixed at Normal GPU accuracy. This results in an up to 40% performance increase, with accurate rendering at normal accuracy
- Red lights in some machines in Xenoblade 2 have been fixed.
- Fire Emblem Warriors has been accurately fixed and no longer requires a workaround.
- Monster Hunter Rise now accurately renders at Normal GPU accuracy. Resulting in an up to 50% performance increase (NOTE: Updates after 3.0.0 still have issues and require more work).
- Vertex Explosions in Persona 5 Royal no longer occur on Normal GPU accuracy. Normal Accuracy provides the best performance by up to 30%
- Atelier Ryza now renders correctly.
- The pessimistic flushes option is no longer needed in any of the affected games.
- Mortal Kombat 11, no longer has any vertex explosions.
- Nier Automata End of Yorha edition now renders correctly.
- Bayonetta 3 no longer requires accuracy high to render correctly.

This has been a very long effort with a lot of back and forth. this was the main reason Y.F.C had to be broken down. It took about 1 year & half to finish this as this is the most crucial part of future improvements coming to yuzu. It's also vital for project Lime.

Special thanks to Maide who helped me figure out the last issues I had on this feature.

Also big thanks to toast for fixing the clang compile errors.

### CODE REVIEW

File: src/video_core/buffer_cache/memory_tracker_base.h [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+            } else if constexpr (create_region_on_fail) {
+                CreateRegion(page_index);
+                manager = top_tier[page_index];
+                execute();
+            }
+            page_index++;
+            page_offset = 0;
+            remaining_size -= copy_amount;
+        }
+        return begin < end ? std::make_pair(begin, end) : std::make_pair(0ULL, 0ULL);
```
Comment (liamwhite) [inline src/video_core/buffer_cache/memory_tracker_base.h:225]:
```suggestion
        if (begin < end) {
            return std::make_pair(begin, end);
        } else {
            return std::make_pair(0ULL, 0ULL);
        }
```
needed for gcc compile

Comment (FernandoS27) [reply src/video_core/buffer_cache/memory_tracker_base.h]: noticed. I did something a bit different.

Comment (liamwhite) [reply src/video_core/buffer_cache/memory_tracker_base.h]: The solution you gave was my first attempt, however it does not work.

Comment (liamwhite) [reply src/video_core/buffer_cache/memory_tracker_base.h]:
```
In file included from src/./video_core/renderer_vulkan/vk_buffer_cache.h:7,
                 from src/video_core/renderer_vulkan/vk_fence_manager.cpp:6:
src/./video_core/buffer_cache/memory_tracker_base.h: In member function ‘std::pair<long unsigned int, long unsigned int> VideoCommon::MemoryTrackerBase<RasterizerInterface>::IteratePairs(VAddr, size_t, Func&&)’:
src/./video_core/buffer_cache/memory_tracker_base.h:225:55: error: cannot bind rvalue reference of type ‘long unsigned int&&’ to lvalue of type ‘u64’ {aka ‘long unsigned int’}
  225 |         return begin < end ? std::make_pair<u64, u64>(begin, end)
      |                                                       ^~~~~
```

File: src/video_core/buffer_cache/memory_tracker_base.h [hunk 2]
```diff
@@ -0,0 +1,266 @@
+// SPDX-FileCopyrightText: Copyright 2022 yuzu Emulator Project
```
Comment (vonchenplus) [inline src/video_core/buffer_cache/memory_tracker_base.h:1]:
```suggestion
// SPDX-FileCopyrightText: Copyright 2023 yuzu Emulator Project
```

File: src/video_core/buffer_cache/word_manager.h [hunk 3]
```diff
@@ -0,0 +1,476 @@
+// SPDX-FileCopyrightText: Copyright 2022 yuzu Emulator Project
```
Comment (vonchenplus) [inline src/video_core/buffer_cache/word_manager.h:1]:
```suggestion
// SPDX-FileCopyrightText: Copyright 2023 yuzu Emulator Project
```

File: src/video_core/renderer_opengl/gl_buffer_cache_base.cpp [hunk 4]
```diff
@@ -0,0 +1,9 @@
+// SPDX-FileCopyrightText: Copyright 2022 yuzu Emulator Project
```
Comment (vonchenplus) [inline src/video_core/renderer_opengl/gl_buffer_cache_base.cpp:1]:
```suggestion
// SPDX-FileCopyrightText: Copyright 2023 yuzu Emulator Project
```

File: src/video_core/renderer_vulkan/vk_buffer_cache_base.cpp [hunk 5]
```diff
@@ -0,0 +1,9 @@
+// SPDX-FileCopyrightText: Copyright 2022 yuzu Emulator Project
```
Comment (vonchenplus) [inline src/video_core/renderer_vulkan/vk_buffer_cache_base.cpp:1]:
```suggestion
// SPDX-FileCopyrightText: Copyright 2023 yuzu Emulator Project
```

**Comment by jacky400** [comment]
@FernandoS27 Just tested it with Mario Spark to see if the shadow issue will be fixed with this buffer cache revamp. Looks like it may be other issue . Any insight what issue it may be ?

image

**Comment by FernandoS27** [comment]
> @FernandoS27 Just tested this PR towards Mario Spark of Hope which persist the rendering issue that think may related to buffer cache. However , the rewrite didn't help and probably other issue . Any insight ?
> 
> image
> 
> image

it's likely a separate issue.

**Comment by piplup55** [comment]
this pr seems to cause a performance regression in botw when using opengl

mainline 1408:
image

Early Access 3524:
image

**Comment by FernandoS27** [comment]
> this pr seems to cause a performance regression in botw when using opengl
> 
> mainline 1408: image
> 
> Early Access 3524: image

will be fixed in the next build.

**Comment by RinMaru** [comment]
This was an issue caused by YFC when it first launched. This build hasnt fixed it unfortunately

File: src/video_core/buffer_cache/buffer_cache.h [hunk 6]
```diff
@@ -1378,27 +995,28 @@ void BufferCache<P>::UpdateIndexBuffer() {
...
     }
     flags[Dirty::IndexBuffer] = false;
-    last_index_count = index_array.count;
     if (!draw_state.inline_index_draw_indexes.empty()) {
         auto inline_index_size = static_cast<u32>(draw_state.inline_index_draw_indexes.size());
         index_buffer = Binding{
             .cpu_addr = 0,
             .size = inline_index_size,
-            .buffer_id = CreateBuffer(0, inline_index_size),
+            .buffer_id = FindBuffer(0, inline_index_size),
```
Comment (vonchenplus) [inline src/video_core/buffer_cache/buffer_cache.h:1009]: if cpu_addr is 0, FindBuffer will return NULL

**Comment by FernandoS27** [comment]
> 
> 
> This was an issue caused by YFC when it first launched. This build hasnt fixed it unfortunately so the issue takes place in sweet mountain act 4 and asteroid coaster act 1 https://drive.google.com/file/d/1f2RoOGZR5-8gxV3DytNefwHQ5fdTrQU6/view?usp=share_link heres my save for easy access

this is unrelated. to this PR.

**Comment by RinMaru** [comment]
> > 
> > This was an issue caused by YFC when it first launched. This build hasnt fixed it unfortunately so the issue takes place in sweet mountain act 4 and asteroid coaster act 1 https://drive.google.com/file/d/1f2RoOGZR5-8gxV3DytNefwHQ5fdTrQU6/view?usp=share_link heres my save for easy access
> 
> this is unrelated. to this PR.

oh sorry. I thought this was finishing YFC and anything it broke at first

File: src/video_core/buffer_cache/buffer_cache.h (hunk 6 above)
Comment (FernandoS27) [reply src/video_core/buffer_cache/buffer_cache.h:1006]: but CreateBuffer continously will keep leaking buffers.

**Review by Morph1984 — APPROVED** [review]

**Comment by Morph1984** [comment]
🏎️

**Comment by coolllman** [comment]
Broke Yoshi's Crafted World

**Comment by Lycoris2013** [comment]
It seems that #10098 is not working properly with Async download when used with #10088. 
I solved the problem by turning it off in Nomal mode like before ed4553806a08e4130fcea36230985cb74d1b326a.
See #10098 for details.
If it is difficult to resolve, can you add a switch to turn it off?

**Pull request merged**